draft-specifications/Model.md

```diff
@@ -0,0 +1,326 @@
+> `Symbol.for('name')` for OCapN selectors, because registered symbols are
+> interned without possibility of eventual garbage collection.
+>
+> OCapN supports one operator for delivering both function application and
+> method invocation.
+> By convention, method invocation is equivalent to function application, where
+> the first argument is a selector followed by the remaining arguments.
+
+# Container
+
+## List (JSON)
+
+A list of any quantity of values.
+
+> - **Guile**: `'()`
+> - **JavaScript** and **JSON**: `[]`
+> - **Python**: `()` (We have not discussed whether to use tuple or list. Tuple
+>   is less mutable, though enforcing immutability in Python is not likely to be
+>   a goal the way it is in hardened JavaScript.)
+>
+> Lists participate in the JSON subset of OCapN.
+>
+> We achieved consensus to name the type "List" at the [November 14, 2023](https://github.com/ocapn/ocapn/issues/94) meeting.
+
+## Struct (JSON)
+
+A struct with unique, unordered string keys and values of heterogeneous type.
+
+> - **Guile**: `make-tbd-hash` a hash of undecided type
+> - **JavaScript**: `{}`
+> - **Python**: `{}`
+>
+> For the purposes of surviving a round trip, the order of appearance of
+> entries in the struct must not be important for determining equivalence.
+> A struct representation concretely using one key order may validly
+> round-trip into a struct representation using another key order.
+> However, we encourage concrete representations to use some canonical key
+> order, though this is not a concern of the abstract syntax and data model.
+>
+> A JavaScript object that owns any non-string keys (numbers or symbols) or
+> with a prototype other than `Object.prototype cannot be passed as an OCapN
+> struct as all such objects are reserved for concrete representations of OCapN
+> types.
+> The key `Symbol.for('passStyle')` is special and indicates the kind of OCapN
+> value the object represents.
+
+## Tagged
+
+A value with a string tag.
+
+> OCapN provides a small number of container types with minimal semantics.
+> Tags allow protocols to emerge from values annotated to indicate a richer
+> variety of semantics.
+> Concretely, a list might be differentiated from or hinted to be an ordered
+> list, sorted list, or a set of unique values.
+>
+> - **Guile**: currently a Guile record labeled `'(syrec ...)` (**imported**)
+> - **JavaScript**: an object with the key `Symbol.for('passStyle')` and value
+>   `tagged`, with a string tagName and a single `"payload"` field carrying the tagged value.
+>   ```js
+>   {
+>     [Symbol.for('passStyle')]: 'tagged',
+>     [Symbol.toStringTag]: tagName,
+>     payload
+>   }
+>   ```
+> - **Python**: `Tagged(tag, value)` where `tag` is a `string` and `Tagged` is
+>   imported from `ocapn`.
+>
+> OCapN tagged data have a single value (are not variadic) [November 14, 2023](https://github.com/ocapn/ocapn/issues/94)
+>
+> What are tagged values for? https://github.com/ocapn/ocapn/issues/52
+
+## Remotable
```

Comment:
@tsyesika @cwebber noted at the October 8, 2024 meeting that we should make this term consistent with the spec, either with spec changes or a change right here. Dave Thompson suggests the corresponding term throughout the spec at time of writing is “remote reference”. Agoric makes a distinction between a remote reference and a “remotable”: a remotable is a settled capability that _may_ be local or remote, not strictly remote.

I think that’s a distinction that applies to this abstract model, and it might be that we need to expand this section to indicate a distinction between a “local presence that forwards messages to a remote reference” and a “a local remotable object that may receive messages from local or remote objects”.

Reply:
The OCapN spec currently simply says "Remote References" which I prefer over "Remoteable" personally.

Reply:
At the November 2024 meeting, we discussed this and resolved that “references” are a category divided between “promise” and “target”, where “targets” have a pass-invariant identity.

We came to this name because the root word “remote” implied that the important distinction between these references and promises is that they can be either “local” or “remote” (or “exported” and “imported” respectively), but that attribute does not distinguish them from promises. The term “target” is intended to capture both of the important distinctions:
- when delivering a message to a reference, the referent may be a promise, but the promise is not the target: it is a conveyance toward the target.
- regardless of whether the target is local or remote, the target has a unique and comparable identity, such that we can establish over OCapN that any pair of references ultimately refer to the same target. We sometimes call this “pass invariant” property of targets “grant matching”. Promises lack this property. In the physical metaphor of “target” a set of concentric circles imply a unique point in space toward which messages (on arrows) converge.